torch/nn/parallel/distributed.py

```diff
@@ -113,14 +113,31 @@ class _DDPUnevenInputsConfig(NamedTuple):
 # is completed.
 class _DDPSink(Function):
     @staticmethod
-    def forward(ctx, reducer, *inputs):
+    def forward(ctx, reducer, state_dict, *inputs):
         ctx.reducer = reducer
+        ctx.state_dict = state_dict
+        ctx.inputs = inputs
         return inputs
 
     @staticmethod
     def backward(ctx, *grad_outputs):
-        Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
-        return (None, *grad_outputs)
+        state_dict = ctx.state_dict
+        static_graph_training = ctx.state_dict['static_graph']
+        if static_graph_training and ctx.state_dict['num_iterations'] == 1:
+            Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
+
+        elif state_dict['find_unused'] and not static_graph_training:
+            # First type of unused params: parameters that did not participate
+            # in computing model outputs. These are found by the below call to
+            # prepare_for_backward.
+            # Second type of unused params: params that won't get gradient
+            # because outputs they produced do not get used in computing loss
+            # for this call to backward. Due to this passthrough autograd
+            # function, autograd hooks for these parameters are now triggered
+            # with zero gradient to maintain parity with local training.
+            ctx.reducer.prepare_for_backward(list(_find_tensors(ctx.inputs)))
```

Comment:
does it mean the unused parameters due to unused output will have zero gradients instead of being 'undefined'? I'm not sure whether this is OK?
also I'm curious about why gradients of unused output tensors are not None..?

Reply:
@zhaojuanmao Yes, if a parameter goes unused because it wasn't used as part of loss computation it will have a zero gradient. This is what the autograd custom function does even on a local model, if it does not receive a grad for the tensor in the backwards pass, it fills it in with zero, see: 

Although, IIUC this shouldn't make a difference? If the grad for the param is k, then we accumulate a grad of `0` onto it and the grad is the same. On the other hand, if grad is `None`, then it will be changed to `0`, but this would happen with `model.zero_grad()` anyways?

Reply:
@zhaojuanmao This should be fixed now with the above ` ctx.set_materialize_grads(False)`.